traffic_ops/traffic_ops_golang/deliveryservice/autorenewcerts.go

```diff
@@ -0,0 +1,190 @@
+			dsExpInfo.XmlId = ds.XmlId
+			dsExpInfo.Version = util.JSONIntStr(int(ds.Version.Int64))
+			dsExpInfo.Error = errors.New("getting ssl keys for xmlId: " + ds.XmlId + " and version: " + strconv.Itoa(int(ds.Version.Int64)) + " :" + err.Error())
+			continue
+		}
+		if !ok {
+			log.Errorf("no object found for the specified key with xmlId: " + ds.XmlId + " and version: " + strconv.Itoa(int(ds.Version.Int64)))
+			dsExpInfo.XmlId = ds.XmlId
+			dsExpInfo.Version = util.JSONIntStr(int(ds.Version.Int64))
+			dsExpInfo.Error = errors.New("no object found for the specified key with xmlId: " + ds.XmlId + " and version: " + strconv.Itoa(int(ds.Version.Int64)))
+			continue
+		}
+
+		err = base64DecodeCertificate(&keyObj.Certificate)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting SSL keys for XMLID '"+ds.XmlId+"': "+err.Error()))
+			return
+		}
+
+		expiration, err := parseExpirationFromCert([]byte(keyObj.Certificate.Crt))
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New(ds.XmlId+": "+err.Error()))
+			return
+		}
+
+		// Renew only certificates within configured limit plus 3 days
+		if expiration.After(time.Now().Add(time.Hour * 24 * time.Duration(inf.Config.ConfigLetsEncrypt.RenewDaysBeforeExpiration)).Add(time.Hour * 24 * 3)) {
+			continue
+		}
+
+		newVersion := util.JSONIntStr(keyObj.Version.ToInt64() + 1)
+
+		dsExpInfo.XmlId = keyObj.DeliveryService
+		dsExpInfo.Version = keyObj.Version
+		dsExpInfo.Expiration = expiration
+		dsExpInfo.AuthType = keyObj.AuthType
+
+		if keyObj.AuthType == tc.LetsEncryptAuthType || (keyObj.AuthType == tc.SelfSignedCertAuthType && inf.Config.ConfigLetsEncrypt.ConvertSelfSigned) {
+			req := tc.DeliveryServiceLetsEncryptSSLKeysReq{
+				DeliveryServiceSSLKeysReq: tc.DeliveryServiceSSLKeysReq{
+					HostName:        &keyObj.Hostname,
+					DeliveryService: &keyObj.DeliveryService,
+					CDN:             &keyObj.CDN,
+					Version:         &newVersion,
+				},
+			}
+			ctx, _ := context.WithTimeout(r.Context(), LetsEncryptTimeout)
+
+			if error := GetLetsEncryptCertificates(inf.Config, req, ctx, inf.User); error != nil {
+				dsExpInfo.Error = error
+			}
+			keysFound.LetsEncryptExpirations = append(keysFound.LetsEncryptExpirations, dsExpInfo)
+
+		} else if keyObj.AuthType == tc.SelfSignedCertAuthType {
+			keysFound.SelfSignedExpirations = append(keysFound.SelfSignedExpirations, dsExpInfo)
+		} else {
+			keysFound.OtherExpirations = append(keysFound.OtherExpirations, dsExpInfo)
+		}
+
+	}
+
+	if inf.Config.SMTP.Enabled && inf.Config.ConfigLetsEncrypt.SendExpEmail {
+		err = AlertExpiringCerts(keysFound, *inf.Config)
+		if err != nil {
+			log.Errorf(err.Error())
+			api.WriteRespAlertObj(w, r, tc.WarnLevel, "Failed to send autorenewal email with error: "+err.Error(), keysFound)
```

Comment:
Because `AlertExpiringCerts` coalesces user errors and system errors to a single error, this can expose messages not meant for the user to see. You should either capture both errors separately or just not tell the user exactly what went wrong.

Reply:
done